### PR TITLE
[#IC-454] Refactored some handlers to be more functional

### PR DESCRIPTION
**Pull request by arcogabbo**
#### List of Changes

* Refactored GetMessage and CreateMessage handlers to be more functional
* Prevent CreateMessage tests from passing cause of missing functions/objects mocks

#### Motivation and Context

fp-ts training while refactoring old code.

#### How Has This Been Tested?

All the tests run fine

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: CreateMessage/handler.ts [hunk 1]
```diff
@@ -441,15 +417,63 @@ export function CreateMessageHandler(
...
+        // The fiscal_code parameter must be specified in the path or in the payload but not in both
+        (payloadFiscalCode: FiscalCode) => pipe(maybeFiscalCodeInPath,
+          O.fold(
+            () => E.right(payloadFiscalCode),
+            () => E.left(ResponseErrorValidation(
+              "Bad parameters",
+              "The fiscalcode parameter must be specified in the path or in the payload but not in both"
+            ))),
+        )
+      ),
```
Comment (AleDore) [inline CreateMessage/handler.ts:451]: Another possible approach:
```
O.foldW(
    () =>
      pipe(
        maybeFiscalCodeInPath,
        E.fromOption(() =>
          ResponseErrorValidation(
            "Bad parameters",
            "The fiscalcode parameter must be specified in the path or in the payload"
          )
        )
      ),
    // The fiscal_code parameter must be specified in the path or in the payload but not in both
    payloadFiscalCode =>
      pipe(
        maybeFiscalCodeInPath,
        E.fromOption(() => payloadFiscalCode),
        E.map(_ =>
          ResponseErrorValidation(
            "Bad parameters",
            "The fiscalcode parameter must be specified in the path or in the payload but not in both"
          )
        ),
        E.swap
      )
  )
```

File: CreateMessage/handler.ts [hunk 2]
```diff
@@ -441,15 +417,63 @@ export function CreateMessageHandler(
...
+    // const maybeFiscalCode = pipe(
+    //   maybeFiscalCodeInPath,
+    //   O.alt(() => maybeFiscalCodeInPayload)
+    // );
+    // if (O.isNone(maybeFiscalCode)) {
+    //   return ResponseErrorValidation(
+    //     "Bad parameters",
+    //     "The fiscalcode parameter must be specified in the path or in the payload"
+    //   );
+    // }
```
Comment (AleDore) [inline CreateMessage/handler.ts:476]: Why this commented code?

File: GetMessage/handler.ts [hunk 3]
```diff
@@ -116,67 +113,31 @@ export function GetMessageHandler(
...
-      retrievedMessage.senderServiceId === userAttributes.service.serviceId ||
-      isUserAllowedForLegalMessages;
 
-    if (!isUserAllowed) {
-      // the user is not allowed to see the message
-      return ResponseErrorForbiddenNotAuthorized;
-    }
+    // the service is allowed to see the message when he is the sender of the message
+    const isUserAllowed = (message: RetrievedMessage) => message.senderServiceId === userAttributes.service.serviceId ||
+      isUserAllowedForLegalMessages ? TE.right(message) : TE.left(ResponseErrorForbiddenNotAuthorized);
```
Comment (AleDore) [inline GetMessage/handler.ts:121]: We can use `TE.fromPredicate` here

File: GetMessage/handler.ts [hunk 4]
```diff
@@ -188,71 +149,82 @@ export function GetMessageHandler(
...
+      // the query failed
+      TE.mapLeft((err) => ResponseErrorQuery(
+        "Error while retrieving the message",
+        err
+      )),
+      // the document does not exist
+      TE.chainW((maybeDocument) => pipe(maybeDocument, TE.fromOption(() => ResponseErrorNotFound(
+        "Message not found",
+        "The message that you requested was not found in the system.")
+      ))),
```
Comment (AleDore) [inline GetMessage/handler.ts:173]: we can safely remove `maybeDocument` explicit definition:
```
TE.chainW(
       TE.fromOption(() => ResponseErrorNotFound(
          "Message not found",
          "The message that you requested was not found in the system.")
        )
)
```

File: GetMessage/handler.ts [hunk 5]
```diff
@@ -188,71 +149,82 @@ export function GetMessageHandler(
...
+      //check if the user is allowed to see the message
+      TE.chainW((document) => isUserAllowed(document)),
```
Comment (AleDore) [inline GetMessage/handler.ts:175]:
```suggestion
      TE.chainW(isUserAllowed),
```

File: GetMessage/handler.ts [hunk 6]
```diff
@@ -188,71 +149,82 @@ export function GetMessageHandler(
...
+            error
+          )}`
+        )),
+        TE.map((maybeMessageStatus) => pipe(maybeMessageStatus,
+          O.map(messageStatus => messageStatus.status),
+          // when the message has been received but a MessageStatus
+          // does not exist yet, the message is considered to be
+          // in the ACCEPTED state (not yet stored in the inbox)
+          O.getOrElse(() => MessageStatusValueEnum.ACCEPTED))
+        ),
```
Comment (AleDore) [inline GetMessage/handler.ts:214]:
```suggestion
        TE.map(flow(
          O.map(messageStatus => messageStatus.status),
          // when the message has been received but a MessageStatus
          // does not exist yet, the message is considered to be
          // in the ACCEPTED state (not yet stored in the inbox)
          O.getOrElse(() => MessageStatusValueEnum.ACCEPTED)
          )
        ),
```

**Review by AleDore — COMMENTED** [review]
nice refactor ;)

File: EmailNotification/handler.ts [hunk 7]
```diff
@@ -163,15 +107,62 @@ export const getEmailNotificationHandler = (
...
-          return EmailNotificationResult.encode({
-            kind: "SUCCESS",
-            result: "OK"
-          });
+          // Check whether the message is expired
+          const errorOrActiveMessage = ActiveMessage.decode(message);
+
+          return await pipe(errorOrActiveMessage,
+            TE.fromEither,
+            TE.foldW(() => {
```
Comment (AleDore) [inline EmailNotification/handler.ts:117]: why not `mapLeft` here and chaining right TaskEither as you fold right?

File: EmailNotification/handler.ts [hunk 8]
```diff
@@ -163,15 +107,62 @@ export const getEmailNotificationHandler = (
...
-          // see #150597597
-          return EmailNotificationResult.encode({
-            kind: "SUCCESS",
-            result: "OK"
-          });
+          // Check whether the message is expired
+          const errorOrActiveMessage = ActiveMessage.decode(message);
+
+          return await pipe(errorOrActiveMessage,
+            TE.fromEither,
```
Comment (AleDore) [inline EmailNotification/handler.ts:116]: we can insert `ActiveMessage` decoding in this pipe

File: EmailNotification/handler.ts [hunk 9]
```diff
@@ -163,15 +107,62 @@ export const getEmailNotificationHandler = (
...
+                  throw new Error(
+                    `Error while fetching the notification: ${JSON.stringify(error)}`
+                  );
+                }),
+                TE.chain(flow(
+                  E.fromOption(() => {
+                    context.log.warn(`${logPrefix}|RESULT=NOTIFICATION_NOT_FOUND`);
+                    throw new Error(`Notification not found`);
+                  }),
+                  E.chain(EmailNotification.decode),
```
Comment (AleDore) [inline EmailNotification/handler.ts:140]: Why decoding here? `notificationModel.find` does not return the correct type?

File: CreateMessage/handler.ts [hunk 10]
```diff
@@ -448,14 +424,41 @@ export function CreateMessageHandler(
...
+          ),
+        ),
+        // The fiscal_code parameter must be specified in the path or in the payload but not in both
+        (payloadFiscalCode: FiscalCode) => pipe(maybeFiscalCodeInPath,
+          E.fromOption(() => payloadFiscalCode),
+          E.map(() => ResponseErrorValidation(
+            "Bad parameters",
+            "The fiscalcode parameter must be specified in the path or in the payload but not in both"
+          )),
+          E.swap,
```
Comment (michaeldisaro) [inline CreateMessage/handler.ts:456]: Very nice!

**Review by michaeldisaro — COMMENTED** [review]
Very very nice use fp-ts. Good job guys!

**Pull request closed without merging**